Common/Scheduling/DateRules.cs

```diff
@@ -220,113 +260,76 @@ private Security GetSecurity(Symbol symbol)
-                        // the date may be before here because we backed up a month
-                        // to properly resolve the first tradeable date
-                        yield return date;
-                    }
-                    lastMonth = date.Month;
-                }
-            }
+            return name;
         }
 
-        private static IEnumerable<DateTime> MonthEndIterator(Security security, DateTime start, DateTime end)
+        private static IEnumerable<DateTime> MonthIterator(Security security, DateTime start, DateTime end, int offset, bool searchForward)
         {
             foreach (var date in Time.EachDay(start, end))
             {
-                if (date.Day == DateTime.DaysInMonth(date.Year, date.Month))
+
+                var daysInMonth = DateTime.DaysInMonth(date.Year, date.Month);
+                var scheduledDayOfMonth = searchForward ?
+                    // For MonthStart we are searching forward so date is 1 + offset; restricted by length of month
+                    1 + offset > daysInMonth ? daysInMonth : 1 + offset
+                    // For MonthEnd we are searching backward so date is daysInMonth - offset; restricted by the 1st
+                    : daysInMonth - offset < 1 ? 1 : daysInMonth - offset;
+
+                // On the day we expect to trigger the event lets find the appropriate day
+                if (date.Day == scheduledDayOfMonth)
                 {
                     if (security == null)
                     {
-                        // fire on the last of each month
+                        // fire on the scheduled day of each month
                         yield return date;
                     }
                     else
                     {
-                        // find previous date when market is open
+                        // find a date when market is open
                         var currentDate = date;
                         while (!security.Exchange.Hours.IsDateOpen(currentDate))
                         {
-                            currentDate = currentDate.AddDays(-1);
+                            // Search in the right direction
+                            currentDate = currentDate.AddDays(searchForward ? 1 : -1);
                         }
                         yield return currentDate;
                     }
                 }
             }
         }
 
-        private static IEnumerable<DateTime> WeekStartIterator(Security security, DateTime start, DateTime end)
+        private static IEnumerable<DateTime> WeekIterator(Security security, DateTime start, DateTime end, int offset, bool searchForward)
         {
-            var skippedMarketClosedDay = false;
+            var scheduledDayOfWeek = searchForward ?
+                // For WeekStart we are searching forward so day of the week is : Monday (1) + offset
+                (DayOfWeek)((1 + offset) % 7) // Math to wrap enum around for Sunday case
```

Comment:
See https://github.com/QuantConnect/Lean/issues/4664, should using Monday, or some other day, as the start of the week depend on the target security/security type? Not sure how if worth the complexity that we would have to add, might get error-prone...
We could also keep the current implementation where we define week start as Monday and week end as Friday, with this `offset` feature it will allow users to achieve whatever weekday they want. Think it's worth adding a test case for `DateRules.WeekStart('EURUSD')` too.
Thoughts @jaredbroad @AlexCatarino @C-SELLERS

Reply:
Yeah now that you mention it it does make sense to have it like that, it would require some refactoring but it should be possible as long as I have access to the "WeekStart" day for any given security. 🤔 

Would we still leave Monday as default "WeekStart" for non-symbol rules?

Reply:
> it would require some refactoring but it should be possible as long as I have access to the "WeekStart" day for any given security. 🤔

yeah, think we could follow something like `start of the week: the day after the last non tradable day of the week`, for always open exchanges it could also default to Monday

> Would we still leave Monday as default "WeekStart" for non-symbol rules?

sounds good 💯